### ISSUE TITLE
[Backend] Rewrite Emitted Code for Closures

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#530]
Implement a new phase `ClosureGen` to introduce a collection of classes for every closure in the program:

For example, given the top-level definition:

```scala
def f(x: Int, y: Bool): Char = ...
```

and the node `Closure(f, List(42))` (formerly known as `MkClosureRef`) emit the class:

```java
final class Clo$f implements Fn1$Bool$Char {
  private int x;
  private boolean y;
  private char result;
  public Clo$f(int x) { this.x = x }
  // methods defined by Fn1$Bool$Char:
  final boolean getArg0() { return this.y; }
  final void setArg0(boolean arg) {  this.y = arg; }
  final char getResult() { return this.result; }
  final void setResult(char result) { this.result = result; }
  final void apply() {
    ... // see note below.
  }
}
```

and emit the following code for the `Closure(f, List(42))` node:

```java
new Clo$f(42);
```

NB: We cannot fully implement `apply` of `Fn1$Bool$Char` without some knowledge of how to retrieve the value of the variables `x` and `y`. Specifically, we need to map variables to some fields, and probably also know whether a variable is closure captured or a regular argument.

**Comment by magnus-madsen** [comment]
Question: Would it be possible to continue using `INVOKEDYNAMIC`?

**Comment by magnus-madsen** [comment]
Fixed #594
